### PR TITLE
fix graphemes eltype for substrings

### LINKED ISSUE #22693: incorrect graphemes eltype for substrings
```jl
julia> s = "123α56789"
"123α56789"

julia> collect(graphemes(SubString(s, 1, 6)))
ERROR: TypeError: copy!: in typeassert, expected SubString{SubString{String}}, got SubString{String}
Stacktrace:
 [1] copy!(::Array{SubString{SubString{String}},1}, ::Base.UTF8proc.GraphemeIterator{SubString{String}}) at ./abstractarray.jl:574
 [2] collect(::Base.UTF8proc.GraphemeIterator{SubString{String}}) at ./array.jl:397

julia> eltype(graphemes(SubString(s, 1, 6)))
SubString{SubString{String}}
```
The `eltype` should be `SubString{String}`, since a substring of a substring of a string is still a `SubString{String}`:
```jl
julia> SubString(SubString(s, 1, 6), 1, 2)
"12"

julia> typeof(ans)
SubString{String}
```

Issue discovered on [discourse](https://discourse.julialang.org/t/regular-expressions-returning-offsets-in-bytes-not-characters/4676/8?u=stevengj), cc @nalimilan.  (Also present in Julia 0.4 and 0.5.)

### PR DESCRIPTION
**Pull request by stevengj**
Fixes #22693.  Also noticed a small simplification in another utf8proc test.

### CODE REVIEW

File: test/unicode/utf8proc.jl [hunk 1]
```diff
@@ -311,9 +311,10 @@ end
...
         @test convert(GenericString, g) == str
-        io = IOBuffer()
-        show(io, g)
-        check = "length-14 GraphemeIterator{String} for \"$str\""
-        @test String(take!(io)) == check
+        @test repr(g) == "length-14 GraphemeIterator{String} for \"$str\""
     end
 end
+
+@testset "graphemes issue #22693" begin
```
Comment (nalimilan) [inline test/unicode/utf8proc.jl]: I would have used a more explicit name like "test grapheme on SubString".

**Review by nalimilan — APPROVED** [review]
Looks good to me, modulo the small remark about the comment.

File: test/unicode/utf8proc.jl (hunk 1 above)
Comment (stevengj) [inline test/unicode/utf8proc.jl]: ok

File: test/unicode/utf8proc.jl [hunk 2]
```diff
@@ -311,9 +311,10 @@ end
...
-        io = IOBuffer()
-        show(io, g)
-        check = "length-14 GraphemeIterator{String} for \"$str\""
-        @test String(take!(io)) == check
+        @test repr(g) == "length-14 GraphemeIterator{String} for \"$str\""
     end
 end
+
+@testset "#22693: substring graphemes" begin
+    @test collect(graphemes(SubString("123α56789", 1, 6))) == ["1","2","3","α","5"]
```
Comment (JeffBezanson) [inline test/unicode/utf8proc.jl]: Also check the eltype?

Comment (stevengj) [inline test/unicode/utf8proc.jl]: ok

**Comment by stevengj** [comment]
AppVeyor failure is #22681.

**Review by StefanKarpinski — APPROVED** [review]

**Comment by stevengj** [comment]
rebased and squashed into two logically independent commits

**Pull request merged**